### DISCUSSION TITLE
아키텍처 초안

### DISCUSSION THREAD

**Discussion by brewagebear** [The-Gleaners/streaming-system#1] [Announcements]
# 데이터플로우 다이어그램 
![image](https://user-images.githubusercontent.com/22961251/233267316-93f71364-0d0c-4ec5-b5fe-728d901f312f.png)

## 상세 설명

1. 이벤트가 발생하면 해당 데이터는 트리거 큐에 인입된다.
2. 이 인입된 이벤트를 다운로더에서 받아서 다운로드를 수행하면서 다운로드 결과값을 지속적으로 다운로드 결과 큐에 보낸다.
3. 프로세서에서는 캐싱된 데이터를 가지고, 신규 혹은 변경데이터를 감지하고 변경될 데이터만 적재 데이터 큐에 보낸다.
4. 로더는 실제로 적재 데이터큐에 쌓인 데이터를 DB에 업로드한다. 

## 프로젝트별 역할 정의
트리거
- 지정한 시간 or 클라이언트 액션에 의하여 요청 메세지 ( Request or Message ) 를 생성하여 다운로더에 전달한다.

다운로더
- 트리거에서 받은 메세지를 컨슈밍한다.
- 메세지에 포함 된 인증 값 및 URL을 이용하여 http 파일 다운로드를 수행한다.
- 다운로드 시작 전 헤더 값에 저장되어 있는 파일 사이즈와 이전 수집 파일 사이즈를 비교하여 수행 여부를 판단한다.
- 파일을 다운로드하며 첫 줄의 헤더를 분석하여 product ( 상품, 리뷰, 여행, 티켓, 기사 등등 ) 를 생성한다.
- 생성 된 product 의 유효성 및 정합성을 검증한다.
- 검증이 완료된 product를 프로세서로 프로듀싱한다.

## 인프라 

상세한 인프라는 아직 정해지지 않았음.
큐를 만약 Kafka로 갈 경우에 데이터 캐싱 부분도 KafkaStreams를 통해서 해결가능해보이나 이 부분은 리서치 필요. 

## 비고 

해당 플로우는 DB에 대한 추가 매핑 및 CDC, 삭제 이벤트는 제외하고 구성되어있음.

**Comment by brewagebear** [comment]
원본 파일은 아래의 링크에서 받으실 수 있습니다.
https://drive.google.com/file/d/1vRvVYhB4K2AjorrwmTrxcevd-O3ccBKk/view?usp=sharing

**Comment by lim-12** [comment]
@brewagebear 님이 작성해주신 내용에 "프로젝트별 역할 정의" 란 추가하였습니다.
제가 맡은 트리거, 다운로더 부분에 대해서는 추가해놨으며, 프로세서와 로더 쪽도 추가 부탁드리겠습니다. ( @suhongkim98 , @jwoo-o , @jongwoolim , @Hyunta )

**Reply by brewagebear** [reply]
@lim-12 님 확인하였습니다. 

내일 정의 추가해두도록 하겠습니다.

**Comment by brewagebear** [comment]
@lim-12 님 인프라 부분 수정하여 위키에 등록하였습니다.
https://github.com/The-Gleaners/streaming-system/wiki/%5BDev%5D-%EC%95%84%ED%82%A4%ED%85%8D%EC%B2%98
